src/pages/library/datasetBuilder/CohortEditor.ts

```diff
@@ -0,0 +1,429 @@
+              h(
+                Link,
+                {
+                  'aria-label': 'delete group',
+                  onClick: () =>
+                    updateCohort(_.set('criteriaGroups', _.without([criteriaGroup], cohort.criteriaGroups))),
+                },
+                [icon('trash')]
+              ),
+            ]),
+          ]
+        ),
+        div([
+          (criteriaGroup.criteria.length !== 0 &&
+            _.map(
+              renderCriteria((criteria: Criteria) =>
+                updateCohort(_.set(`criteriaGroups.${index}.criteria`, _.without([criteria], criteriaGroup.criteria)))
+              ),
+              criteriaGroup.criteria
+            )) ||
+            div({ style: { marginTop: 15 } }, [
+              div({ style: { fontWeight: 'bold', fontStyle: 'italic' } }, ['No criteria yet']),
+              div({ style: { fontStyle: 'italic', marginTop: 5 } }, [
+                "You can add a criteria by clicking on 'Add criteria'",
+              ]),
+            ]),
+        ]),
+        div({ style: { marginTop: 10 } }, [
+          h(Select, {
+            styles: { container: (provided) => ({ ...provided, width: '230px' }) },
+            isClearable: false,
+            isSearchable: false,
+            options: [
+              {
+                label: 'Domains',
+                options: _.map((domainType) => {
+                  return {
+                    value: domainType,
+                    label: domainType.category,
+                  };
+                }, datasetDetails.domainTypes),
+              },
+              {
+                label: 'Program Data',
+                options: _.map((programDataType) => {
+                  return {
+                    value: programDataType,
+                    label: programDataType.name,
+                  };
+                }, datasetDetails.programDataTypes),
+              },
+            ],
+            'aria-label': 'add criteria',
+            placeholder: 'Add criteria',
+            value: undefined,
+            onChange: (x) => {
+              // FIXME: is there a way to remove any?
+              const criteria = createCriteriaFromType((x as any).value);
+              updateCohort(_.set(`criteriaGroups.${index}.criteria.${criteriaGroup.criteria.length}`, criteria));
+            },
+          }),
+        ]),
+      ]),
+      div(
+        {
+          style: {
+            paddingTop: 5,
+            marginTop: 10,
+            marginBottom: 10,
+            borderTop: `1px solid ${colors.dark(0.35)}`,
+            display: 'flex',
+            justifyContent: 'flex-end',
+            fontWeight: 'bold',
+          },
+        },
+        [div({ style: { marginRight: 10 } }, [`Group count: ${criteriaGroup.count}`])]
+      ),
+    ]
+  );
+};
+
+const RenderCohort = ({
+  datasetDetails,
+  cohort,
+  updateCohort,
+}: {
+  cohort: Cohort | undefined;
```

Comment:
this type information is far more readable if defined before the component as XComponentProps interface.

Also, RenderCohort feels like a wierd naming... why is the render verb leading the name of a component?

Reply:
OK, I can add a props type to all component methods in this file and rename this to `CohortView`.

Reply:
If it's up to our team, I think I'll leave it as-is, although I think the name `render` is misleading here so I'll rename it.

> But at a minimum, methods like that should be defined OUTSIDE the React.FC's function body, since that's the render loop and should be mindful of performance

In this case the method is defined outside the component body, so I think this is OK. Did you see something here that needs to be addressed?